#16541310 Validate user registration

ericnyirimana opened the pull request:
Description
=======
Every system application needs to validate inserted data by the user. This pull request is for user signup data validation which consists of validating data user object and provide a descriptive response error message.

Type of change
=======
N/A

How has it been tested
=======
- It is tested when you signup via ```api/v1/users/signup```

Checklist:
=======
N/A

Pivotal tracker story ID
=======
[16541310](https://www.pivotaltracker.com/n/projects/2326984/stories/16541310)

ericnyirimana committed:
feat(validation): Validate user registration
- Validate username
- Validate email
- Validate password
- Wrote unit test
- Documented the feature
[WIP #165413101]

AJAkimana reviewed (commented):
I suggest that it will be great if you declare those regular expressions apart, then you use variables.

ericnyirimana committed:
Feat(validation): Remove validation from controllers
- Created a middleware for signup validation
- Refactor an article test password to meet middleware
[WIP #165413101]

macphilips commented on controllers/user.controller.js:
```diff
@@ -76,13 +76,11 @@ class UserController {
-      if (err) {
-        res.status(409).send({ status: 409, message: err.parent.detail });
-      }
+      return res.status(500).send({ status: 500, message: 'Internal Server Error' });
```
What is the reason for return internal server error?

ericnyirimana replied:
I have validated all expected error from this endpoint, So else I'm expecting HTTP status 500 for bad scripting error ( internal server error )
